python/gstaichi/lang/ast/ast_transformer_utils.py

```diff
@@ -265,22 +280,35 @@ def check_loop_var(self, loop_var: str) -> None:
                 f"Variable '{loop_var}' is already declared in the outer scope and cannot be used as loop variable"
             )
 
-    def get_var_by_name(self, name: str) -> Any:
+    def get_var_by_name(self, name: str) -> tuple[PtrSource, Any]:
         for s in reversed(self.local_scopes):
             if name in s:
-                return s[name]
-        if name in self.global_vars:
+                val = s[name]
+                ptr_source = PtrSource.LOCAL
+                if self.is_pure and is_data_oriented(val):
+                    # a data oriented is as good as global, from pov of pure constraints
+                    ptr_source = PtrSource.GLOBAL
```

Comment:
This sounds a bit hacky no? And the error message in the end may be confusing?

Reply:
I'm open to renaming `global` to something else if you wish. `Non-pure` perhaps?

Reply:
Cool. One think I noticed is that when we initialize this, often we don't know whether the node is pure compliant or not. If we initialize with `is_pure_compliant=True`, then it makes it sound like we are sure that it is compliant, which might not be the case. So I've flipped it around to`violates_pure=False`. And then, if we discovert hat it does violate pure, we can flip this to `True`, for a node

Reply:
The thing is, we never now when its determined. It basically defaults to False, until we find a violation.

Think of it like a flag 'is_criminal' on a human. If someone is convicted of a crime, then, assuming they were correctly convicted, we know we can set this flag to True. But otherwise, we don't know if it's true or false.

I suppose what we could do is have only two possible states:
- `None`
- `True`

Reply:
(or another analogy:when running pytest, a flag like `exist_failing_tests`. Initially False, later might become True).